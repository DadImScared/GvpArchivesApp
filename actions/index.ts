export { default as autoComplete } from "./autoComplete";
export { default as audioPlayer } from "./audioPlayer";
export { default as itemsByCategory } from "./itemsByCategory";
export { default as itemsById } from "./itemsById";
export { default as loading } from "./loading";
export { default as search } from "./search";
