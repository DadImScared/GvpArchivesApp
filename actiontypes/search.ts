export const UPDATE_SEARCH_QUERY = "search/UPDATE_SEARCH_QUERY";
export const TOGGLE_CATEGORIES = "search/TOGGLE_CATEGORIES";
export const SELECT_ALL_CATEGORIES = "search/SELECT_ALL_CATEGORIES";
export const REMOVE_ALL_CATEGORIES = "search/REMOVE_ALL_CATEGORIES";
