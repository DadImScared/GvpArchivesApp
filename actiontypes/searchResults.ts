export const ADD_SEARCH = "searchResults/ADD_SEARCH";
export const UPDATE_SEARCH = "searchResults/UPDATE_SEARCH";
