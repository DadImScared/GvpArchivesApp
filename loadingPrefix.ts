export const ITEMS_BY_CATEGORY = "ITEMS_BY_CATEGORY_";
export const AUTOCOMPLETE_QUERY = "AUTOCOMPLETE_QUERY_";
export const SEARCH_RESULTS = "SEARCH_RESULTS_";
