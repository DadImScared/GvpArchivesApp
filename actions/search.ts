import axios, { AxiosPromise } from "axios";
import { AnyAction, Dispatch } from "redux";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/search";
import { AUTOCOMPLETE_QUERY } from "../loadingPrefix";
import { IReducerState } from "../reducers";
import { getQueryId } from "../reducers/search";
import autoComplete from "./autoComplete";
import { BASE_URL } from "./itemsById";

const autoCompleteUrl = `${BASE_URL}/api/v1/completeme/`;

export const isValidValidTimeToLive = (time: number) => time ? time >= +new Date() : false;

const updateSearchQuery = (query: string) => action(ActionTypes.UPDATE_SEARCH_QUERY, query);

type ApiCallCallback<ReturnType> = (
  state: IReducerState,
  dispatch: Dispatch
) => ReturnType;

type IAfterApiCall<Data> = (
  data: Data,
  state: IReducerState,
  dispatch: Dispatch
) => Promise<AnyAction[]> | AnyAction[];

export interface IApiCallConfig<Data> {
  afterApiCall?: IAfterApiCall<Data>;
  apiCall: ApiCallCallback<AxiosPromise<any>>;
  beforeApiCall?: ApiCallCallback<Promise<AnyAction[]> | AnyAction[]>;
  loadingId?: ((state: IReducerState) => string) | string;
  shouldCallApi?: ApiCallCallback<boolean | AxiosPromise<void>>;
}

export const getQuerySuggestions = (query: string, categories: string[]): IApiCallConfig<string[]> => {
  const queryId = getQueryId(query, categories);
  const loadingId = `${AUTOCOMPLETE_QUERY}${queryId}`;
  return {
    loadingId,
    afterApiCall: (data) => {
      return [
        autoComplete.addAutoComplete(
          query,
          categories,
          data.map((textObj: any) => textObj.text)
        )
      ];
    },
    apiCall: () => axios.get(`${autoCompleteUrl}${query}`),
    shouldCallApi: (state) => {
      const isLoading = state.loading[loadingId];
      if (!query || isLoading) return false;

      const currentAutoComplete = state.autoComplete[queryId];
      if (!currentAutoComplete) return true;

      return isValidValidTimeToLive(currentAutoComplete.timeToLive) ?
        false
        :
        axios.get(`${autoCompleteUrl}${query}`);
    }
  };
};

export default {
  updateSearchQuery
};
