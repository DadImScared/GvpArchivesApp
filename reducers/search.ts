import _ from "lodash";
import { ActionType } from "typesafe-actions";

import { search } from "../actions";
import * as ActionTypes from "../actiontypes/search";
import { IReducerState } from "./index";

export interface ISearchState {
  categories: string[];
  query: string;
}

export type SearchStateAndActions<
  T extends keyof ISearchState | keyof typeof search
  > = Pick<ISearchState & typeof search, T>;

export const CATEGORIES = [
  "bhagavatpatrika",
  "book",
  "harikatha",
  "harmonistmagazine",
  "harmonistmonthly",
  "lecture",
  "movie",
  "song"
];

export const getInitialSearchState = (): ISearchState => ({
  categories: CATEGORIES,
  query: ""
});

export default function(state = getInitialSearchState(), action: ActionType<typeof search>) {
  switch (action.type) {
    case ActionTypes.UPDATE_SEARCH_QUERY:
      return {
        ...state,
        query: action.payload
      };
    case ActionTypes.TOGGLE_CATEGORIES:
      const { payload } = action;
      const { categories } = state;
      return {
        ...state,
        categories: categories.includes(payload) ?
          categories.filter((category) => category !== action.payload)
          :
          [...categories, payload]
      };
    case ActionTypes.SELECT_ALL_CATEGORIES:
      return {
        ...state,
        categories: [...CATEGORIES]
      };
    case ActionTypes.REMOVE_ALL_CATEGORIES:
      return {
        ...state,
        categories: []
      };
    default:
      return state;
  }
}

export const getQueryId = (query: string, categories: string[]) => {
  return `${query}-${_.sortBy(categories, (o) => o).join(",")}`;
};

export const getSearch = (state: IReducerState) => state.search;
