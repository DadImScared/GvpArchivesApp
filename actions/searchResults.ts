import axios, { AxiosResponse } from "axios";
import { action } from "typesafe-actions";

import * as ActionTypes from "../actiontypes/searchResults";
import { SEARCH_RESULTS } from "../loadingPrefix";
import { getCategoriesWithDefault } from "../reducers/autoComplete";
import { Item } from "../reducers/itemsById";
import { getQueryId } from "../reducers/search";
import { makeQueryString } from "../utils/queryString";
import { addItemsById, BASE_URL } from "./itemsById";
import { IApiCallConfig, isValidValidTimeToLive } from "./search";

const SEARCH_URL = `${BASE_URL}/api/v1/search`;

export interface ISearchItem extends Item {
  highlightedTitle?: string;
  highlightedBody?: string;
}

interface ISearchResults {
  nextPage: string | boolean;
  results: ISearchItem[];
  suggestions: Array<{ text: string }>;
}

export interface INormalizedSearchResults { item_id: string; highlightedBody?: string; highlightedTitle?: string; }

const addSearch = (
  queryId: string,
  results: INormalizedSearchResults[],
  suggestions: string[],
  nextPage: boolean | string
) => action(ActionTypes.ADD_SEARCH, { queryId, results, suggestions, nextPage });

const updateSearch = (
  queryId: string,
  results: INormalizedSearchResults[],
  suggestions: string[],
  nextPage: boolean | string
) => action(ActionTypes.UPDATE_SEARCH, { queryId, results, suggestions, nextPage });

export const normalizeSearchResults = (items: ISearchItem[]) => {
  const itemsById: { [key: string]: Item } = {};
  const normalizedValues: INormalizedSearchResults[] = [];
  items.forEach(({highlightedBody, highlightedTitle, ...item }) => {
    normalizedValues.push({
      item_id: item.item_id,
      [highlightedBody ? "highlightedBody" : "highlightedTitle"]: highlightedBody || highlightedTitle
    });
    itemsById[item.item_id] = item;
  });
  return {
    itemsById,
    results: normalizedValues
  };
};

export const getSearchResults = (
  query: string,
  selectedCategories: string[],
  refresh = false
): IApiCallConfig<AxiosResponse<ISearchResults>> => {
  const categories = getCategoriesWithDefault(selectedCategories);
  const queryId = getQueryId(query, categories);
  const loadingId = `${SEARCH_RESULTS}${queryId}`;
  const firstPage = `${SEARCH_URL}/${query}/?${makeQueryString({ categories })}`;
  let shouldUpdateSearch = false;
  return {
    loadingId,
    afterApiCall: ({ data }) => {
      // const { result: itemIds, entities: { items } } = normalize(data.results, itemSchema);
      const { itemsById, results } = normalizeSearchResults(data.results);
      return [
        addItemsById(itemsById),
        // add or update search based on bool 'shouldUpdateSearch'
        (shouldUpdateSearch ? updateSearch : addSearch)(
          queryId,
          results,
          data.suggestions.map((suggestion) => suggestion.text),
          data.nextPage
        )
      ];
    },
    apiCall: (state) => axios.get(firstPage),
    shouldCallApi: (state) => {
      if (refresh) return true;
      const isLoading = state.loading[loadingId];

      if (!query || isLoading) return false;

      const currentSearch = state.searchResults[queryId];
      if (!currentSearch) return true;

      if (isValidValidTimeToLive(currentSearch.timeToLive)) {
        if (currentSearch.nextPage) {
          shouldUpdateSearch = true;
          return axios.get(`${BASE_URL}${currentSearch.nextPage}`);
        }
        return false;
      }

      return true;
    },
    type: "callApiMiddleware"
  };
};

export default {
  addSearch,
  updateSearch
};
