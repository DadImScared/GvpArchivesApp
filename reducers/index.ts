import _ from "lodash";
import { combineReducers, DeepPartial } from "redux";

import audioPlayer, { getInitialAudioPlayerState, IAudioPlayerState } from "./audioPlayer";
import autoComplete, { getInitialAutoCompleteState, IAutoCompleteState } from "./autoComplete";
import itemsByCategory, { ItemsByCategoryState } from "./itemsByCategory";
import itemsById, { ItemsByIdState } from "./itemsById";
import loading, { ILoadingState } from "./loading";
import recentSearches, { RecentSearches } from "./recentSearches";
import search, { getInitialSearchState, ISearchState } from "./search";
import searchResults, { ISearchResults } from "./searchResults";

export interface IReducerState {
  audioPlayer: IAudioPlayerState;
  autoComplete: IAutoCompleteState;
  loading: ILoadingState;
  itemsByCategory: ItemsByCategoryState;
  itemsById: ItemsByIdState;
  recentSearches: RecentSearches;
  search: ISearchState;
  searchResults: ISearchResults;
}

export const getInitialReducerState = (injectedState: DeepPartial<IReducerState> = {}) => _.merge({
  audioPlayer: getInitialAudioPlayerState(),
  autoComplete: getInitialAutoCompleteState(),
  itemsByCategory: {},
  itemsById: {},
  loading: {},
  recentSearches: [],
  search: getInitialSearchState(),
  searchResults: {}
}, injectedState);

const reducer = combineReducers<IReducerState>({
  audioPlayer,
  autoComplete,
  itemsByCategory,
  itemsById,
  loading,
  recentSearches,
  search,
  searchResults
});

export default reducer;
