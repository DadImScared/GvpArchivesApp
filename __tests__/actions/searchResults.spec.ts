import axios from "axios";
jest.mock("axios");

import { getSearchResults, normalizeSearchResults } from "../../actions/searchResults";
import { SEARCH_RESULTS } from "../../loadingPrefix";
import { getInitialReducerState } from "../../reducers";
import { getQueryId } from "../../reducers/search";
import { makeQueryString } from "../../utils/queryString";

describe("searchResults", () => {
  describe("normalizeSearchResults", () => {
    it("should normalize results", () => {
      const items = [
        {
          category: "harikatha",
          highlightedBody: "[In 2007, in honor of Sri <em>Radhastami</em>, Srila Bhaktivedanta Narayana Gosvami Maharaja gave two discourses on the glories of Srimati Radhika; one on the eve of <em>Radhastami</em> and one on the following morning.",
          item_id: "78d1a9dc-3785-46f1-a6db-407e1fb40c2d",
          link: "http://www.purebhakti.com/teachers/bhakti-discourses-mainmenu-61/52-discourses-2009/1086-the-glory-of-sri-radha.html",
          title: "The Glory of Sri Radha",
        },
        {
          category: "harikatha",
          highlightedTitle: "Here is  a commentary, given on Sri <em>Radhastami</em> Day, two weeks later.  Here again, by the manifestation of the mercy of Srimati Radhika in the heart of Srila Maharaja, ever-new meanings are revealed.]",
          item_id: "ae955019-8cb3-4f4d-a578-48d9c065847d",
          link: "http://www.purebhakti.com/teachers/bhakti-discourses-mainmenu-61/22-discourses-2003/367-sri-radhastami-day.html",
          title: "Sri Radhastami Day"
        }
      ];
      expect(normalizeSearchResults(items)).toMatchSnapshot();
    });
  });
  describe("getSearchResults", () => {

    beforeEach(() => {
      jest.resetAllMocks();
    });

    const successResponse = {
      data: {
        nextPage: false,
        results: [{ item_id: "item1", category: "book", title: "item-1", link: "item-1-link" }],
        suggestions: []
      }
    };
    const query = "query";
    const categories = ["book", "movie"];

    it("should have a loading ID", () => {
      const results = getSearchResults(query, categories);
      expect(results.loadingId).toEqual(`${SEARCH_RESULTS}${getQueryId(query, categories)}`);
    });

    it("should not fetch data if loading", () => {
      const results = getSearchResults(query, categories);
      (axios.get as any).mockImplementationOnce(() => Promise.resolve(successResponse));
      const loading = {
        [`${SEARCH_RESULTS}${getQueryId(query, categories)}`]: true
      };
      expect(
        results.shouldCallApi && results.shouldCallApi(getInitialReducerState({
          loading: {
            ...loading
          }
        }), jest.fn())
      ).toEqual(false);
    });

    it("should have correct first page url and add search", async () => {
      (axios.get as any).mockImplementationOnce((url: string) => {
        expect(url.split(".com")[1]).toEqual(
          `/api/v1/search/query/?${makeQueryString({ categories })}`
        );
        return Promise.resolve(successResponse);
      });
      const results = getSearchResults(query, categories);
      expect(
        results.afterApiCall && results.afterApiCall(
          await results.apiCall(getInitialReducerState(), jest.fn()),
          getInitialReducerState(),
          jest.fn()
        )
      ).toMatchSnapshot();
    });

    it("should have correct next page url and update search", async () => {
      (axios.get as any).mockImplementationOnce((url: string) => {
        expect(url.split(".com")[1]).toEqual(
          `/api/v1/search/query/?${makeQueryString({ categories, page: 2 })}`
        );
        return Promise.resolve(successResponse);
      });
      const state = getInitialReducerState(
        {
          searchResults: {
            [getQueryId(query, categories)]: {
              nextPage: `/api/v1/search/query/?${makeQueryString({ categories, page: 2 })}`,
              results: [],
              suggestions: [],
              timeToLive: (+new Date()) + 2000
            }
          }
        }
      );
      const dispatch = jest.fn();
      const results = getSearchResults(query, categories);
      const apiResults: any = await (results.shouldCallApi && results.shouldCallApi(state, dispatch));
      const afterApiCall = results.afterApiCall && results.afterApiCall(apiResults, state, dispatch);
      expect(afterApiCall).toMatchSnapshot();
    });
  });
});
