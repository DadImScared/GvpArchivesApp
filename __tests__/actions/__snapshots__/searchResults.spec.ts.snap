// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searchResults getSearchResults should have correct first page url and add search 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "items": Object {
        "item1": Object {
          "category": "book",
          "item_id": "item1",
          "link": "item-1-link",
          "title": "item-1",
        },
      },
    },
    "type": "itemsById/ADD_ITEMS",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "nextPage": false,
      "queryId": "query-book,movie",
      "results": Array [
        Object {
          "highlightedTitle": undefined,
          "item_id": "item1",
        },
      ],
      "suggestions": Array [],
    },
    "type": "searchResults/ADD_SEARCH",
  },
]
`;

exports[`searchResults getSearchResults should have correct next page url and update search 1`] = `
Array [
  Object {
    "meta": undefined,
    "payload": Object {
      "items": Object {
        "item1": Object {
          "category": "book",
          "item_id": "item1",
          "link": "item-1-link",
          "title": "item-1",
        },
      },
    },
    "type": "itemsById/ADD_ITEMS",
  },
  Object {
    "meta": undefined,
    "payload": Object {
      "nextPage": false,
      "queryId": "query-book,movie",
      "results": Array [
        Object {
          "highlightedTitle": undefined,
          "item_id": "item1",
        },
      ],
      "suggestions": Array [],
    },
    "type": "searchResults/UPDATE_SEARCH",
  },
]
`;

exports[`searchResults normalizeSearchResults should normalize results 1`] = `
Object {
  "itemsById": Object {
    "78d1a9dc-3785-46f1-a6db-407e1fb40c2d": Object {
      "category": "harikatha",
      "item_id": "78d1a9dc-3785-46f1-a6db-407e1fb40c2d",
      "link": "http://www.purebhakti.com/teachers/bhakti-discourses-mainmenu-61/52-discourses-2009/1086-the-glory-of-sri-radha.html",
      "title": "The Glory of Sri Radha",
    },
    "ae955019-8cb3-4f4d-a578-48d9c065847d": Object {
      "category": "harikatha",
      "item_id": "ae955019-8cb3-4f4d-a578-48d9c065847d",
      "link": "http://www.purebhakti.com/teachers/bhakti-discourses-mainmenu-61/22-discourses-2003/367-sri-radhastami-day.html",
      "title": "Sri Radhastami Day",
    },
  },
  "results": Array [
    Object {
      "highlightedBody": "[In 2007, in honor of Sri <em>Radhastami</em>, Srila Bhaktivedanta Narayana Gosvami Maharaja gave two discourses on the glories of Srimati Radhika; one on the eve of <em>Radhastami</em> and one on the following morning.",
      "item_id": "78d1a9dc-3785-46f1-a6db-407e1fb40c2d",
    },
    Object {
      "highlightedTitle": "Here is  a commentary, given on Sri <em>Radhastami</em> Day, two weeks later.  Here again, by the manifestation of the mercy of Srimati Radhika in the heart of Srila Maharaja, ever-new meanings are revealed.]",
      "item_id": "ae955019-8cb3-4f4d-a578-48d9c065847d",
    },
  ],
}
`;
